Compiler/src/Transform.hs

```diff
@@ -0,0 +1,159 @@
+    transformedK <- transformLocator k
+    pure $ Select transformedL transformedK
+
+transformTransformer :: (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => Transformer a -> State (Env c) (Transformer b)
+transformTransformer (Construct name args) = do
+    transformedArgs <- mapM transformLocator args
+    pure $ Construct name transformedArgs
+
+transformTransformer (Call name args) = do
+    transformedArgs <- mapM transformLocator args
+    pure $ Call name transformedArgs
+
+transformPrecondition :: (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => Precondition a -> State (Env c) (Precondition b)
+transformPrecondition (Conj conds) = do
+    transformedConds <- mapM transformPrecondition conds
+    pure $ Conj transformedConds
+
+transformPrecondition (Disj conds) = do
+    transformedConds <- mapM transformPrecondition conds
+    pure $ Disj transformedConds
+
+transformPrecondition (BinOp op a b) = do
+    transformedA <- transformLocator a
+    transformedB <- transformLocator b
+    pure $ BinOp op transformedA transformedB
+
+transformPrecondition (NegateCond cond) = do
+    transformedCond <- transformPrecondition cond
+    pure $ NegateCond transformedCond
+
+transformStmt :: (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => Stmt a -> State (Env c) (Stmt b)
+transformStmt (Flow src dst) = do
+    transformedSrc <- transformLocator src
+    transformedDst <- transformLocator dst
+    pure $ Flow transformedSrc transformedDst
+
+transformStmt (FlowTransform src transformer dst) = do
+    transformedSrc <- transformLocator src
+    transformedDst <- transformLocator dst
+    transformedTransformer <- transformTransformer transformer
+    pure $ FlowTransform transformedSrc transformedTransformer transformedDst
+
+transformStmt (OnlyWhen precondition) = do
+    transformedPrecondition <- transformPrecondition precondition
+    pure $ OnlyWhen transformedPrecondition
+
+transformStmt (Try checkC1 checkCs) = do
+    transformedC1 <- mapM transformStmt checkC1
+    transformedCs <- mapM transformStmt checkCs
+    pure $ Try transformedC1 transformedCs
+
+transformStmt Revert = pure Revert
+
+transformDecl :: (Phase a, Phase b, Phase c, ProgramTransform a b, PhaseTransition a c) => Decl a -> State (Env c) (Decl b)
+transformDecl (TypeDecl s modifiers baseT) = do
+    transformedBaseT <- transformBaseType baseT
+    pure $ TypeDecl s modifiers transformedBaseT
+
+transformDecl (TransformerDecl name args ret body) = do
+    transformedArgs <- mapM transformVarDef args
+    transformedRet <- transformVarDef ret
+    transformedBody <- mapM transformStmt body
+    pure $ TransformerDecl name transformedArgs transformedRet transformedBody
+
+transformEnv :: forall a b. (Phase a, Phase b, ProgramTransform a b, PhaseTransition a a) => Env a -> Env b
+transformEnv env = let (transformedTypeEnvs, s) = runState (mapM transformBaseType (view typeEnv env)) env
```

Comment:
Not a big deal, but probably should be consistent between using `view` and `(^.)`